File: pkg/taskstate/store.go
package taskstate

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/nats-io/nats.go/jetstream"
)

const (
	// BucketTasks holds serialized TaskState values. Key = task ID.
	BucketTasks = "wasm-af-tasks"
	// BucketAudit is an append-only audit log. Keys are
	// "<task-id>.<unix-nano>" to maintain ordering within a task.
	BucketAudit = "wasm-af-audit"
	// BucketPayloads holds step input/output payloads separately from task
	// state to avoid bloating the state KV entry. Key = step output key.
	BucketPayloads = "wasm-af-payloads"
)

// Store provides task state operations backed by NATS JetStream KV.
type Store struct {
	tasks    jetstream.KeyValue
	audit    jetstream.KeyValue
	payloads jetstream.KeyValue
}

// NewStore creates or updates the three JetStream KV buckets and returns a Store.
// It is safe to call repeatedly — existing buckets are re-used via CreateOrUpdateKeyValue.
func NewStore(ctx context.Context, js jetstream.JetStream) (*Store, error) {
	tasks, err := js.CreateOrUpdateKeyValue(ctx, jetstream.KeyValueConfig{
		Bucket:      BucketTasks,
		Description: "wasm-af task states",
		History:     10, // retain 10 revisions per task for CAS
	})
	if err != nil {
		return nil, fmt.Errorf("tasks bucket: %w", err)
	}

	audit, err := js.CreateOrUpdateKeyValue(ctx, jetstream.KeyValueConfig{
		Bucket:       BucketAudit,
		Description:  "wasm-af immutable audit log",
		History:      1,        // each key is unique; no need for history
		MaxValueSize: 64 << 10, // 64 KB per audit entry
	})
	if err != nil {
		return nil, fmt.Errorf("audit bucket: %w", err)
	}

	payloads, err := js.CreateOrUpdateKeyValue(ctx, jetstream.KeyValueConfig{
		Bucket:       BucketPayloads,
		Description:  "wasm-af step input/output payloads",
		History:      1,
		MaxValueSize: 4 << 20, // 4 MB per payload
	})
	if err != nil {
		return nil, fmt.Errorf("payloads bucket: %w", err)
	}

	return &Store{tasks: tasks, audit: audit, payloads: payloads}, nil
}

// Put writes a TaskState. The state's UpdatedAt is always set to now.
func (s *Store) Put(ctx context.Context, state *TaskState) error {
	state.UpdatedAt = time.Now().UTC()
	b, err := json.Marshal(state)
	if err != nil {
		return fmt.Errorf("marshal: %w", err)
	}
	_, err = s.tasks.Put(ctx, state.TaskID, b)
	return err
}

// Get retrieves the current TaskState for the given task ID.
func (s *Store) Get(ctx context.Context, taskID string) (*TaskState, error) {
	entry, err := s.tasks.Get(ctx, taskID)
	if err != nil {
		return nil, fmt.Errorf("kv get %s: %w", taskID, err)
	}
	var state TaskState
	if err := json.Unmarshal(entry.Value(), &state); err != nil {
		return nil, fmt.Errorf("unmarshal: %w", err)
	}
	return &state, nil
}

// Update performs a read-modify-write with optimistic CAS.
// fn is called with a mutable copy of the current state. If it returns an error,
// Update aborts without writing. If a concurrent writer wins the CAS race,
// Update retries up to 5 times before returning an error.
func (s *Store) Update(ctx context.Context, taskID string, fn func(*TaskState) error) error {
	const maxRetries = 5

	for i := range maxRetries {
		entry, err := s.tasks.Get(ctx, taskID)
		if err != nil {
			return fmt.Errorf("cas read (attempt %d): %w", i+1, err)
		}

		var state TaskState
		if err := json.Unmarshal(entry.Value(), &state); err != nil {
			return fmt.Errorf("unmarshal: %w", err)
		}

		if err := fn(&state); err != nil {
			return err // caller signalled abort
		}

		state.UpdatedAt = time.Now().UTC()
		b, err := json.Marshal(&state)
		if err != nil {
			return fmt.Errorf("marshal: %w", err)
		}

		_, err = s.tasks.Update(ctx, taskID, b, entry.Revision())
		if err == nil {
			return nil
		}
		if !errors.Is(err, jetstream.ErrKeyExists) {
			return fmt.Errorf("cas write: %w", err)
		}
		// CAS conflict — retry
	}
	return fmt.Errorf("too many CAS conflicts for task %s", taskID)
}

// Delete removes a task state entry.
func (s *Store) Delete(ctx context.Context, taskID string) error {
	return s.tasks.Delete(ctx, taskID)
}

// AppendAudit writes an immutable audit event.
// The key is "<task-id>.<unix-nano>" to guarantee ordering per task.
func (s *Store) AppendAudit(ctx context.Context, event *AuditEvent) error {
	if event.Timestamp.IsZero() {
		event.Timestamp = time.Now().UTC()
	}
	b, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("marshal audit: %w", err)
	}
	key := fmt.Sprintf("%s.%d", event.TaskID, event.Timestamp.UnixNano())
	_, err = s.audit.Put(ctx, key, b)
	return err
}

// PutPayload stores a step payload under the given key.
func (s *Store) PutPayload(ctx context.Context, key, payload string) error {
	_, err := s.payloads.Put(ctx, key, []byte(payload))
	return err
}

// GetPayload retrieves a step payload by key.
func (s *Store) GetPayload(ctx context.Context, key string) (string, error) {
	entry, err := s.payloads.Get(ctx, key)
	if err != nil {
		return "", fmt.Errorf("payload get %s: %w", key, err)
	}
	return string(entry.Value()), nil
}

// DeletePayload removes a payload entry after the step has been processed.
func (s *Store) DeletePayload(ctx context.Context, key string) error {
	return s.payloads.Delete(ctx, key)
}

File: crates/taskstate/src/store.rs
use crate::types::{AuditEvent, TaskState};
use async_nats::jetstream;
use chrono::Utc;
use std::sync::atomic::{AtomicU64, Ordering};
use thiserror::Error;
use tracing::warn;

const BUCKET_TASKS: &str = "wasm-af-tasks";
const BUCKET_AUDIT: &str = "wasm-af-audit";
const BUCKET_PAYLOADS: &str = "wasm-af-payloads";

const MAX_CAS_RETRIES: u32 = 5;

#[derive(Debug, Error)]
pub enum StoreError {
    #[error("nats: {0}")]
    Nats(String),
    #[error("serialization: {0}")]
    Serde(#[from] serde_json::Error),
    #[error("not found: {0}")]
    NotFound(String),
    #[error("CAS conflict after {0} retries for task {1}")]
    CasConflict(u32, String),
    #[error("update aborted: {0}")]
    UpdateAborted(String),
}

impl From<async_nats::Error> for StoreError {
    fn from(e: async_nats::Error) -> Self {
        StoreError::Nats(e.to_string())
    }
}

pub struct Store {
    tasks: jetstream::kv::Store,
    audit: jetstream::kv::Store,
    payloads: jetstream::kv::Store,
    audit_seq: AtomicU64,
}

impl Store {
    pub async fn new(js: jetstream::Context) -> Result<Self, StoreError> {
        let tasks = js
            .create_key_value(jetstream::kv::Config {
                bucket: BUCKET_TASKS.to_string(),
                description: "wasm-af task states".to_string(),
                history: 10,
                ..Default::default()
            })
            .await
            .map_err(|e| StoreError::Nats(format!("tasks bucket: {e}")))?;

        let audit = js
            .create_key_value(jetstream::kv::Config {
                bucket: BUCKET_AUDIT.to_string(),
                description: "wasm-af immutable audit log".to_string(),
                history: 1,
                max_value_size: 64 * 1024,
                ..Default::default()
            })
            .await
            .map_err(|e| StoreError::Nats(format!("audit bucket: {e}")))?;

        let payloads = js
            .create_key_value(jetstream::kv::Config {
                bucket: BUCKET_PAYLOADS.to_string(),
                description: "wasm-af step input/output payloads".to_string(),
                history: 1,
                max_value_size: 4 * 1024 * 1024,
                ..Default::default()
            })
            .await
            .map_err(|e| StoreError::Nats(format!("payloads bucket: {e}")))?;

        Ok(Store {
            tasks,
            audit,
            payloads,
            audit_seq: AtomicU64::new(0),
        })
    }

    pub async fn put(&self, state: &mut TaskState) -> Result<(), StoreError> {
        state.updated_at = Utc::now();
        let bytes = serde_json::to_vec(state)?;
        self.tasks
            .put(&state.task_id, bytes.into())
            .await
            .map_err(|e| StoreError::Nats(format!("put {}: {e}", state.task_id)))?;
        Ok(())
    }

    pub async fn get(&self, task_id: &str) -> Result<TaskState, StoreError> {
        let entry = self
            .tasks
            .entry(task_id)
            .await
            .map_err(|e| StoreError::Nats(format!("get {task_id}: {e}")))?;
        match entry {
            Some(entry) => {
                let state: TaskState = serde_json::from_slice(&entry.value)?;
                Ok(state)
            }
            None => Err(StoreError::NotFound(task_id.to_string())),
        }
    }

    /// Read-modify-write with optimistic CAS. Retries up to MAX_CAS_RETRIES
    /// on revision conflicts.
    pub async fn update<F>(&self, task_id: &str, f: F) -> Result<(), StoreError>
    where
        F: Fn(&mut TaskState) -> Result<(), String>,
    {
        for attempt in 0..MAX_CAS_RETRIES {
            let entry = self
                .tasks
                .entry(task_id)
                .await
                .map_err(|e| StoreError::Nats(format!("cas read (attempt {}): {e}", attempt + 1)))?;

            let entry = match entry {
                Some(e) => e,
                None => return Err(StoreError::NotFound(task_id.to_string())),
            };

            let revision = entry.revision;
            let mut state: TaskState = serde_json::from_slice(&entry.value)?;

            if let Err(msg) = f(&mut state) {
                return Err(StoreError::UpdateAborted(msg));
            }

            state.updated_at = Utc::now();
            let bytes = serde_json::to_vec(&state)?;

            match self.tasks.update(task_id, bytes.into(), revision).await {
                Ok(_) => return Ok(()),
                Err(e) => {
                    let err_str = e.to_string();
                    if err_str.contains("wrong last sequence") {
                        warn!(task_id, attempt, "CAS conflict, retrying");
                        continue;
                    }
                    return Err(StoreError::Nats(format!("cas write: {err_str}")));
                }
            }
        }
        Err(StoreError::CasConflict(MAX_CAS_RETRIES, task_id.to_string()))
    }

    pub async fn delete(&self, task_id: &str) -> Result<(), StoreError> {
        self.tasks
            .delete(task_id)
            .await
            .map_err(|e| StoreError::Nats(format!("delete {task_id}: {e}")))?;
        Ok(())
    }

    pub async fn append_audit(&self, event: &mut AuditEvent) -> Result<(), StoreError> {
        if event.timestamp == chrono::DateTime::UNIX_EPOCH {
            event.timestamp = Utc::now();
        }
        let bytes = serde_json::to_vec(event)?;
        let seq = self.audit_seq.fetch_add(1, Ordering::Relaxed);
        let key = format!(
            "{}.{}.{}",
            event.task_id,
            event.timestamp.timestamp_nanos_opt().unwrap_or(0),
            seq
        );
        self.audit
            .put(&key, bytes.into())
            .await
            .map_err(|e| StoreError::Nats(format!("audit put: {e}")))?;
        Ok(())
    }

    pub async fn put_payload(&self, key: &str, payload: &str) -> Result<(), StoreError> {
        let bytes: bytes::Bytes = payload.as_bytes().to_vec().into();
        self.payloads
            .put(key, bytes)
            .await
            .map_err(|e| StoreError::Nats(format!("payload put {key}: {e}")))?;
        Ok(())
    }

    pub async fn get_payload(&self, key: &str) -> Result<String, StoreError> {
        let entry = self
            .payloads
            .entry(key)
            .await
            .map_err(|e| StoreError::Nats(format!("payload get {key}: {e}")))?;
        match entry {
            Some(e) => Ok(String::from_utf8_lossy(&e.value).to_string()),
            None => Err(StoreError::NotFound(key.to_string())),
        }
    }

    pub async fn delete_payload(&self, key: &str) -> Result<(), StoreError> {
        self.payloads
            .delete(key)
            .await
            .map_err(|e| StoreError::Nats(format!("payload delete {key}: {e}")))?;
        Ok(())
    }

    pub fn tasks_kv(&self) -> &jetstream::kv::Store {
        &self.tasks
    }
}
